python/ray/train/trainer.py

```diff
@@ -603,19 +633,74 @@ def _run_with_error_handling(self, func: Callable):
     def __next__(self):
         if self.is_finished():
             raise StopIteration
-        next_results = self._run_with_error_handling(
-            self._executor.fetch_next_result)
+        next_results = self._run_with_error_handling(self._fetch_next_result)
         if next_results is None:
             try:
-                self._final_results = \
-                    self._run_with_error_handling(
-                        self._executor.finish_training)
+                self._final_results = self._run_with_error_handling(
+                    self._finish_training)
             finally:
                 self._finished_training = True
             raise StopIteration
         else:
+
             return next_results
 
+    def _fetch_next_result(self) -> Optional[List[Dict]]:
+        """Fetch next results produced by ``train.report()`` from each worker.
+
+        Assumes ``start_training`` has already been called.
+
+        Returns:
+            A list of dictionaries of values passed to ``train.report()`` from
+                each worker. Each item corresponds to an intermediate result
+                a single worker. If there are no more items to fetch,
+                returns None.
+        """
+
+        while True:
+            results = ray.get(self._executor.get_next_results.remote())
+            if results is None:
+                return None
+            first_result = results[0]
+            result_type = first_result.type
+            if result_type is TrainingResultType.REPORT:
+                result_data = [r.data for r in results]
+                return result_data
+            elif result_type is TrainingResultType.CHECKPOINT:
+                self._checkpoint_manager._process_checkpoint(results)
+                # Iterate until next REPORT call or training has finished.
+            else:
+                raise TrainBackendError(f"Unexpected result type: "
+                                        f"{result_type}. "
+                                        f"Expected one of "
+                                        f"{[type in TrainingResultType]}")
+
+    def _finish_checkpointing(self):
```

Comment:
Ah, the reason we had this differentiation in the first place was because there might be 1 report still enqueued, so we wouldn't be able to just call `_fetch_next_result`.

Reply:
Ok got it. I think we can still do some refactoring to remove this, but let's leave that for a future PR.